File: cours/examples/example-serialization-and-communication/src/main/java/ch/heigvd/amt/protobuf/Protobuf.java
package ch.heigvd.amt.protobuf;

import ch.heigvd.amt.proto.PersonOuterClass;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;

public class Protobuf {

    public static void main(String[] args) throws Exception {
        // Create the persons
        var p1 = PersonOuterClass.Person.newBuilder()
                .setName("John")
                .setAge(42)
                .build();
        var p2 = PersonOuterClass.Person.newBuilder()
                .setName("Jane")
                .setAge(36)
                .build();
        var persons = PersonOuterClass.Persons.newBuilder()
                .addPersons(p1)
                .addPersons(p2)
                .build();

        // Set the file name
        var file = new File("persons.pb");

        // Serialize to binary
        try (var fos = new FileOutputStream(file)) {
            persons.writeTo(fos);
        }

        // Deserialize from binary
        try (var fis = new FileInputStream(file)) {
            var deserializedWrapper = PersonOuterClass.Persons.parseFrom(fis);
            deserializedWrapper.getPersonsList().forEach(person -> System.out.println(person.getName() + " " + person.getAge()));
        }
    }

}

File: cours/examples/example-serialization-and-communication/src/main/java/ch/heigvd/amt/protobuf/Protobuf.js
// Intall the protoc compiler
// brew install grpc protobuf

// Install the protoc-gen-js plugin
// npm install protoc-gen-js -g

// Install the google-protobuf library
// npm install google-protobuf --save

// Install the protobufjs library
// npm install protobufjs --save

// Generate the JS file
// Note: The path of js_out (.) is relative to the proto file (person.proto)
// protoc --js_out=import_style=commonjs,binary:. src/main/resources/proto/person.proto

// Import the generated JS file
const personProto = require('../../../../../resources/proto/person_pb');

// Read the binary data from a file
const fs = require('fs');
const binaryData = fs.readFileSync('persons.pb');

// Deserialize the binary data
const persons = personProto.Persons.deserializeBinary(binaryData);

// Display the persons
persons.getPersonsList().forEach(person => {
    console.log(person.getName());
    console.log(person.getAge());
});

// Run the script
// node src/main/java/ch/heigvd/amt/protobuf/Protobuf.js
